### PR TITLE
[CI] Benchmark pipeline

### PR DESCRIPTION
**Pull request by kaisecheng**
Add a buildkite pipeline to do benchmark.
The script does benchmark by running Filebeats (docker) -> Logstash (docker) -> ES Cloud.
Logstash metrics and benchmark results are sent to the same ES Cloud.
- Secrets store in vault `secret/ci/elastic-logstash/benchmark`
- Use flog (docker) to generate ~2GB logs
- Pull the snapshot docker image of the main branch every day
- Logstash runs two pipelines, main and node_stats
  - The main pipeline handles beats ingestion, sending data to the data stream `logs-generic-default`
    - It runs for all combinations. (pq + mq) x worker x batch size
    - Each test runs for ~7 minutes
  - The node_stats pipeline retrieves /_node/stats API every 30s and sends it to the data stream `metrics-nodestats-logstash`
- The script sends a summary of EPS and resource usage to index `benchmark_summary`

The buildkite pipeline accepts ENV variables to customize the test
| Variable Name   | Default Value       | Comment                                       |
|-----------------|---------------------|-----------------------------------------------|
| FB_VERSION      | 8.13.4              | docker tag                                    |
| LS_VERSION      |     | docker tag                                    |
| LS_JAVA_OPTS    | -Xmx2g              | by default, Xmx is set to half of memory      |
| MULTIPLIERS     | 2,4,6               | determine the number of workers (cpu * multiplier) |
| BATCH_SIZES     | 125,1000            |                                               |
| CPU             | 4                   | number of cpu for Logstash container          |
| MEM             | 4                   | number of GB for Logstash container           |
| QTYPE           | memory              | queue type to test -- persisted; memory; all     |
| FB_CNT          | 4                   | number of filebeats to use in benchmark       |

To check the result
- `vault read secret/ci/elastic-logstash/benchmark` to get the host and credentials
- `curl -u "$ES_USER:$ES_PW" "$ES_HOST/benchmark_summary/_search"`

### CODE REVIEW

**Comment by kaisecheng** [comment]
build https://buildkite.com/elastic/logstash-benchmark-pipeline/builds/20#018fd13a-ddd1-4bff-9249-6d81316f9ee4

**Comment by kaisecheng** [comment]
buildkite: https://buildkite.com/elastic/logstash-benchmark-pipeline/builds/35#01900785-efc2-4608-b109-cff1777bfc18

File: .buildkite/scripts/benchmark/main.sh [hunk 1]
```diff
@@ -1,9 +1,316 @@
...
+  pull_images
+  check_logs
+
+  NS_DIR="fb${FB_CNT}c${CPU}m${MEM}" # fb4c4m4
+  mkdir -p "$SCRIPT_PATH/$NS_DIR"
+
+  if [[ $QTYPE == "all" ]]; then
+    queue
+  else
+    worker
```
Comment (mashhurs) [inline .buildkite/scripts/benchmark/main.sh:309]: I would rather absorb these lines (306L~310L) into `run_pipeline` to improve the logic readability.

**Review by mashhurs — COMMENTED** [review]
This is great work @kaisecheng !
I have just placing my high level review and I will leave shell script to @dliappis.

- daily run:  do we _really_ need daily basis benchmark? It seems to me weekly collection is also okay.
- what is the end goal? LS performance monitoring? We are not including default plugins (only beats and es-output) in the config and I am a bit confused about our end goal.

File: .buildkite/scripts/benchmark/main.sh [hunk 2]
```diff
@@ -1,9 +1,316 @@
...
+  MAX_WORKER_CONCURR=$( jqmax '.pipelines.main.flow.worker_concurrency.last_1_minute' "$file_glob" )
+  MAX_Q_EVENT_CNT=$( jqmax '.pipelines.main.queue.events_count' "$file_glob" )
+  MAX_Q_SIZE=$( jqmax '.pipelines.main.queue.queue_size_in_bytes' "$file_glob" )
+
+  AVG_CPU_PERCENT=$( jqavg '.process.cpu.percent' "$file_glob" )
+  AVG_HEAP=$( jqavg '.jvm.mem.heap_used_in_bytes' "$file_glob" )
+  AVG_NON_HEAP=$( jqavg '.jvm.mem.non_heap_used_in_bytes' "$file_glob" )
+}
+
+send_summary() {
```
Comment (mashhurs) [inline .buildkite/scripts/benchmark/main.sh:209]: one more thing: I do see sometimes we have an empty data. Can we filter out them?
Example doc:
```
{
                "_index": "benchmark_summary",
                "_id": "jNa22Y8BWqGBxUhZpwMn",
                "_score": 1.0,
                "_source":
                {
                    "timestamp": "2024-06-02T16:09:37",
                    "version": "",
                    "cpu": "",
                    "mem": "",
                    "workers": "",
                    "batch_size": "",
                    "queue_type": "",
                    "total_events_out": "",
                    "max_eps_1m": "",
                    "max_eps_5m": "",
                    "max_worker_utilization": "",
                    "max_worker_concurrency": "",
                    "avg_cpu_percentage": "",
                    "avg_heap": "",
                    "avg_non_heap": "",
                    "max_queue_events": "",
                    "max_queue_bytes_size": ""
                }
            },
```

**Comment by kaisecheng** [comment]
@mashhurs thanks for the review
The goal of doing daily ingestion is to see the trend of eps. When we see the eps is dropping, we can trace back which commits cause the issue. This is aiming to test logstash core and not individual plugin, but we can extend the scope to include the most popular plugins later.

With this tool, we can do benchmark for all versions. Sometimes, SDHs come with a message saying 7.17 has higher throughput and use less resources. I hope this daily benchmark gives us confident that the performance is on the same bar and we should actively trace the benchmark instead of waiting customer to report it.

File: .buildkite/scripts/benchmark/main.sh (hunk 2 above)
Comment (kaisecheng) [reply .buildkite/scripts/benchmark/main.sh:209]: That is leftover of old testing data. Let me delete them.

File: .buildkite/scripts/benchmark/main.sh [hunk 3]
```diff
@@ -1,9 +1,316 @@
...
+    IMAGE_URL="https://snapshots.elastic.co/${BUILD_ID}/downloads/logstash/logstash-$LS_VERSION-docker-image-$ARCH.tar.gz"
+    IMAGE_FILENAME="$LS_VERSION.tar.gz"
+
+    echo "Download $LS_VERSION from $IMAGE_URL"
+    [[ ! -e $IMAGE_FILENAME ]] && curl -fsSL --retry-max-time 60 --retry 3 --retry-delay 5 -o "$IMAGE_FILENAME" "$IMAGE_URL"
+    [[ -z $(docker images -q docker.elastic.co/logstash/logstash:$LS_VERSION) ]] && docker load -i "$IMAGE_FILENAME"
+  fi
+
+  # pull filebeat image
+  FB_LATEST_VERSION=$(curl --retry-all-errors --retry 5 --retry-delay 1 -s "https://api.github.com/repos/elastic/beats/tags" | jq -r '.[0].name' | cut -c 2-)
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply .buildkite/scripts/benchmark/main.sh:121]: I wanna use stable version or even a fix version for filebeat and elasticsearch to exclude the performance fluctuation and make sure we are measuring the logstash core

File: .buildkite/scripts/benchmark/main.sh [hunk 4]
```diff
@@ -1,9 +1,316 @@
...
+  read -ra BATCH_SIZES <<< "$BATCH_SIZES"
+
+  IFS=' '
+  echo "filebeats: $FB_CNT, cpu: $CPU, mem: $MEM, Queue: $QTYPE, worker multiplier: ${MULTIPLIERS[@]}, batch size: ${BATCH_SIZES[@]}"
+}
+
+get_secret() {
+  VAULT_PATH=secret/ci/elastic-logstash/benchmark
+  VAULT_DATA=$(vault kv get -format json $VAULT_PATH)
+  BENCHMARK_ES_HOST=$(echo $VAULT_DATA | jq -r '.data.es_host')
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply .buildkite/scripts/benchmark/main.sh:93]: It is ES cloud 8.13.3. Logstash support wide range of ES version. At this moment, version compatibility is not the concern.

File: .buildkite/scripts/benchmark/main.sh [hunk 5]
```diff
@@ -1,9 +1,316 @@
...
+aggregate_stats() {
+  local file_glob="$SCRIPT_PATH/$NS_DIR/${QTYPE:0:1}_w${WORKER}b${BATCH_SIZE}_*.json"
+  MAX_EPS_1M=$( jqmax '.flow.output_throughput.last_1_minute' "$file_glob" )
+  MAX_EPS_5M=$( jqmax '.flow.output_throughput.last_5_minutes' "$file_glob" )
+  MAX_WORKER_UTIL=$( jqmax '.pipelines.main.flow.worker_utilization.last_1_minute' "$file_glob" )
+  MAX_WORKER_CONCURR=$( jqmax '.pipelines.main.flow.worker_concurrency.last_1_minute' "$file_glob" )
+  MAX_Q_EVENT_CNT=$( jqmax '.pipelines.main.queue.events_count' "$file_glob" )
+  MAX_Q_SIZE=$( jqmax '.pipelines.main.queue.queue_size_in_bytes' "$file_glob" )
+
+  AVG_CPU_PERCENT=$( jqavg '.process.cpu.percent' "$file_glob" )
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply .buildkite/scripts/benchmark/main.sh:204]: 👍

**Comment by kaisecheng** [comment]
@mashhurs :) great to get your feedback. When I wrote the last comment, `elastic_package` 💡 in my head. Only know a little bit about this amazing project. If you can point me to the how-to-setup elastic_package, I can prepare a python version for the next onWeek. IMO, getting the benchmark to collect metrics from the main branch today is better than delaying for the next improved version. 

For (1), I think we won't need to run benchmark regularly on old versions. My idea is to manually trigger buildkite job for v7.17 and all minor versions of v8. Hopefully, it will be enough to plot a graph. And I think it is unlikely to have performance change in v7.17, as neither jruby nor java version will upgrade. 

For (2), This single buildkite pipeline won't achieve all goals. I expect to have another pipeline to collect benchmark for the new released versions. For "include last and current releases weekly", current and last releases are branched from main, and critical fixes will be merged to main and backported to minor version. If I create the branch 8.14 today and schedule to release it next month, I would expect monitoring 8.15-snapshot performance is enough. Tracing the last branch, in this case 8.14 bugfixes, is optional and I am opened to add it in another pipeline

File: .buildkite/scripts/benchmark/main.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaisecheng) [reply .buildkite/scripts/benchmark/main.sh:93]: Thinking it twice, I think testing against local ES docker and allowing customized ES version give more flexibility. Multiple benchmark jobs can run parallelly without affecting the ingest eps of ES Cloud. I think if the GCP instance is big enough and disk IOPs is enough, putting everything in the same instance should be fine.

**Review by dliappis — APPROVED** [review]
Thanks for adding this!

I left a comment about general methodology. I have similar thoughts to @mashhurs , I think it would be trivial to rewrite this to a language like Python to increase readability / maintainability etc.

**Pull request merged**